internal/client.go

```diff
@@ -431,50 +432,69 @@ func (c *Client) OnRequestCompleted(rc RequestCompletionCallback) {
 	c.onRequestCompleted = rc
 }
 
+func (c *Client) clearCookies() {
+	c.cookies = nil
+}
+
 // Do performs request passed
 func (c *Client) Do(ctx context.Context, req *http.Request, v interface{}) error {
 	// check if httpClient exists or not
+	maxRetries := 1
+	currentTry := 0
+
 	if c.httpClient == nil {
 		return fmt.Errorf(c.ErrorMsg)
 	}
 
-	req = req.WithContext(ctx)
-	resp, err := c.httpClient.Do(req)
-	if err != nil {
-		return err
-	}
-
-	defer func() {
-		if rerr := resp.Body.Close(); err == nil {
-			err = rerr
+	for currentTry <= maxRetries {
+		req = req.WithContext(ctx)
+		resp, err := c.httpClient.Do(req)
+		if err != nil {
+			return err
 		}
-	}()
-
-	err = CheckResponse(resp)
 
-	if err != nil {
-		return err
-	}
+		defer func() {
+			if rerr := resp.Body.Close(); err == nil {
+				err = rerr
+			}
+		}()
 
-	if v != nil {
-		if w, ok := v.(io.Writer); ok {
-			_, err = io.Copy(w, resp.Body)
-			if err != nil {
-				fmt.Printf("Error io.Copy %s", err)
-				return err
+		if err := CheckResponse(resp); err != nil {
+			if resp.StatusCode == http.StatusUnauthorized && c.credentials.SessionAuth && currentTry < maxRetries {
+				currentTry++
```

Comment:
cleaner way to add a retry functionality is by passing a function to retry to a retryFunc external function https://medium.com/@reetas/different-ways-to-retry-function-execution-in-golang-763229550dab

Reply:
I did some minor refactoring to have a recursive `do` call. We can do the decorator pattern as mentioned but I don't think that needs to happen in this PR.